airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlCdcTargetPosition.java

```diff
@@ -0,0 +1,104 @@
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package io.airbyte.integrations.source.mssql;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.google.common.base.Preconditions;
+import io.airbyte.db.jdbc.JdbcDatabase;
+import io.airbyte.db.jdbc.JdbcUtils;
+import io.airbyte.integrations.debezium.CdcTargetPosition;
+import io.airbyte.integrations.debezium.internals.SnapshotMetadata;
+import io.debezium.connector.sqlserver.Lsn;
+import java.io.IOException;
+import java.sql.SQLException;
+import java.util.List;
+import java.util.Optional;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class MssqlCdcTargetPosition implements CdcTargetPosition {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(MssqlCdcTargetPosition.class);
+  private final Lsn targetLsn;
+
+  public MssqlCdcTargetPosition(Lsn targetLsn) {
+    this.targetLsn = targetLsn;
+  }
+
+  @Override
+  public boolean reachedTargetPosition(JsonNode valueAsJson) {
+    Lsn recordLsn = extractLsn(valueAsJson);
+
+    if (targetLsn.compareTo(recordLsn) > 0) {
+      return false;
+    } else {
+      SnapshotMetadata snapshotMetadata = SnapshotMetadata.valueOf(valueAsJson.get("source").get("snapshot").asText().toUpperCase());
+      // if not snapshot or is snapshot but last record in snapshot.
+      return SnapshotMetadata.TRUE != snapshotMetadata;
+    }
+  }
+
+  private Lsn extractLsn(JsonNode valueAsJson) {
+    return Optional.ofNullable(valueAsJson.get("source"))
+        .flatMap(source -> Optional.ofNullable(source.get("commit_lsn").asText()))
+        .map(Lsn::valueOf)
+        .orElseThrow(() -> new IllegalStateException("Could not find LSN"));
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+    MssqlCdcTargetPosition that = (MssqlCdcTargetPosition) o;
+    return targetLsn.equals(that.targetLsn);
+  }
+
+  @Override
+  public int hashCode() {
+    return targetLsn.hashCode();
+  }
+
+  public static MssqlCdcTargetPosition getTargetPostion(JdbcDatabase database) {
```

Comment:
would it make sense to write unit tests for this?

Reply:
Addressed with a basic test in https://github.com/airbytehq/airbyte/pull/4689/commits/2f87d62186ad5bd99b496afe4a0529b9b0e099cb

Possibly worth writing unit tests for whole class but bad implementation should cause the CdcMssqlSourceTests to fail anyway so not sure it's urgent